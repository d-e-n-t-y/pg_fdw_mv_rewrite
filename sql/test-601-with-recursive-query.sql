-- Anticipate message "query requires unsupported features"
set mv_rewrite.log_match_progress = 'true';

explain (VERBOSE, COSTS OFF)
with recursive data as (
    select 1 n, key, value
      from test
      where key = 'key1'
      and value = 1::text
  union all
    select curr.n + 1, next.key, next.value
      from data curr, test next
      where
        curr.key = next.key
        and next.value = (curr.value::numeric + 1)::text
)
select key, count (value)
  from data
  group by key;

with recursive data as (
    select 1 n, key, value
      from test
      where key = 'key1'
      and value = 1::text
  union all
    select curr.n + 1, next.key, next.value
      from data curr, test next
      where
        curr.key = next.key
        and next.value = (curr.value::numeric + 1)::text
)
select key, count (value)
  from data
  group by key;
