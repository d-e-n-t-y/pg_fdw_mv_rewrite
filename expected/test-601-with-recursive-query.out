-- Anticipate message "query requires unsupported features"
set mv_rewrite.log_match_progress = 'true';
explain (VERBOSE, COSTS OFF)
with recursive data as (
    select 1 n, key, value
      from test
      where key = 'key1'
      and value = 1::text
  union all
    select curr.n + 1, next.key, next.value
      from data curr, test next
      where
        curr.key = next.key
        and next.value = (curr.value::numeric + 1)::text
)
select key, count (value)
  from data
  group by key;
INFO:  mv_rewrite_add_rewritten_mv_paths: evaluating MV: public.test_mv1
INFO:  mv_rewrite_check_group_clauses_for_mv: looking to rewrite JOIN, but MV is GROUP BY.
INFO:  mv_rewrite_add_rewritten_mv_paths: evaluating MV: public.test_mv2
INFO:  mv_rewrite_check_group_clauses_for_mv: looking to rewrite JOIN, but MV is GROUP BY.
INFO:  mv_rewrite_add_rewritten_mv_paths: evaluating MV: public.test_expr_mv1
INFO:  mv_rewrite_check_group_clauses_for_mv: looking to rewrite JOIN, but MV is GROUP BY.
INFO:  mv_rewrite_add_rewritten_mv_paths: evaluating MV: public.test_union_mv1
INFO:  mv_rewrite_join_node_is_valid_for_plan: the query does not involve all relations joined by the MV
INFO:  mv_rewrite_add_rewritten_mv_paths: evaluating MV: public.test_union_mv2
INFO:  mv_rewrite_join_node_is_valid_for_plan: the query does not involve all relations joined by the MV
INFO:  mv_rewrite_add_rewritten_mv_paths: evaluating MV: public.test_mv1
INFO:  mv_rewrite_check_group_clauses_for_mv: looking to rewrite JOIN, but MV is GROUP BY.
INFO:  mv_rewrite_add_rewritten_mv_paths: evaluating MV: public.test_mv2
INFO:  mv_rewrite_check_group_clauses_for_mv: looking to rewrite JOIN, but MV is GROUP BY.
INFO:  mv_rewrite_add_rewritten_mv_paths: evaluating MV: public.test_expr_mv1
INFO:  mv_rewrite_check_group_clauses_for_mv: looking to rewrite JOIN, but MV is GROUP BY.
INFO:  mv_rewrite_add_rewritten_mv_paths: evaluating MV: public.test_union_mv1
INFO:  mv_rewrite_join_node_is_valid_for_plan: the query does not involve all relations joined by the MV
INFO:  mv_rewrite_add_rewritten_mv_paths: evaluating MV: public.test_union_mv2
INFO:  mv_rewrite_join_node_is_valid_for_plan: the query does not involve all relations joined by the MV
INFO:  mv_rewrite_create_upper_paths_hook: query requires unsupported features.
                                                      QUERY PLAN                                                      
----------------------------------------------------------------------------------------------------------------------
 HashAggregate
   Output: data.key, count(data.value)
   Group Key: data.key
   CTE data
     ->  Recursive Union
           ->  Gather
                 Output: 1, test.key, test.value
                 Workers Planned: 2
                 ->  Parallel Seq Scan on public.test
                       Output: 1, test.key, test.value
                       Filter: ((test.key = 'key1'::text) AND (test.value = '1'::text))
           ->  Hash Join
                 Output: (curr.n + 1), next.key, next.value
                 Hash Cond: ((next.key = curr.key) AND (next.value = (((curr.value)::numeric + '1'::numeric))::text))
                 ->  Seq Scan on public.test next
                       Output: next.key, next.hidden, next.value
                 ->  Hash
                       Output: curr.n, curr.key, curr.value
                       ->  WorkTable Scan on data curr
                             Output: curr.n, curr.key, curr.value
   ->  CTE Scan on data
         Output: data.n, data.key, data.value
(22 rows)

with recursive data as (
    select 1 n, key, value
      from test
      where key = 'key1'
      and value = 1::text
  union all
    select curr.n + 1, next.key, next.value
      from data curr, test next
      where
        curr.key = next.key
        and next.value = (curr.value::numeric + 1)::text
)
select key, count (value)
  from data
  group by key;
INFO:  mv_rewrite_add_rewritten_mv_paths: evaluating MV: public.test_mv1
INFO:  mv_rewrite_check_group_clauses_for_mv: looking to rewrite JOIN, but MV is GROUP BY.
INFO:  mv_rewrite_add_rewritten_mv_paths: evaluating MV: public.test_mv2
INFO:  mv_rewrite_check_group_clauses_for_mv: looking to rewrite JOIN, but MV is GROUP BY.
INFO:  mv_rewrite_add_rewritten_mv_paths: evaluating MV: public.test_expr_mv1
INFO:  mv_rewrite_check_group_clauses_for_mv: looking to rewrite JOIN, but MV is GROUP BY.
INFO:  mv_rewrite_add_rewritten_mv_paths: evaluating MV: public.test_union_mv1
INFO:  mv_rewrite_join_node_is_valid_for_plan: the query does not involve all relations joined by the MV
INFO:  mv_rewrite_add_rewritten_mv_paths: evaluating MV: public.test_union_mv2
INFO:  mv_rewrite_join_node_is_valid_for_plan: the query does not involve all relations joined by the MV
INFO:  mv_rewrite_add_rewritten_mv_paths: evaluating MV: public.test_mv1
INFO:  mv_rewrite_check_group_clauses_for_mv: looking to rewrite JOIN, but MV is GROUP BY.
INFO:  mv_rewrite_add_rewritten_mv_paths: evaluating MV: public.test_mv2
INFO:  mv_rewrite_check_group_clauses_for_mv: looking to rewrite JOIN, but MV is GROUP BY.
INFO:  mv_rewrite_add_rewritten_mv_paths: evaluating MV: public.test_expr_mv1
INFO:  mv_rewrite_check_group_clauses_for_mv: looking to rewrite JOIN, but MV is GROUP BY.
INFO:  mv_rewrite_add_rewritten_mv_paths: evaluating MV: public.test_union_mv1
INFO:  mv_rewrite_join_node_is_valid_for_plan: the query does not involve all relations joined by the MV
INFO:  mv_rewrite_add_rewritten_mv_paths: evaluating MV: public.test_union_mv2
INFO:  mv_rewrite_join_node_is_valid_for_plan: the query does not involve all relations joined by the MV
INFO:  mv_rewrite_create_upper_paths_hook: query requires unsupported features.
 key  | count 
------+-------
 key1 |    10
(1 row)

